### ISSUE TITLE
Wishlist - make audio stream available via stdout

### ISSUE THREAD

**Issue by malfino** [Opendigitalradio/dablin#4]
dablin is the only program with which I was able to get DAB+ running. I'd like to stream one single radio-stream on my lan. As dablin decodes the data stream and directly sends it to sdl output, there doesn't seem to be a way to transport the stream  on the network. 

It'd be great, if there could be a cmd line switch to get the audio stream on stdout.

I'd like to use something like this:

```
$ ./dablin -d ../dabtools/src/dab2eti -c 7D -s 0xD332 -d - | vlc -vvv stream:///dev/stdin --sout '#rtp{sdp=rtsp://:8554/}'
```

**Comment by basicmaster** [comment]
It would be quite easy to output the not yet decoded audio stream to stdout, but there is a problem regarding DAB+ services:
DAB+ uses AAC with 960 samples/frame instead of the default 1024 samples/frame - and the target decoder must get to know this someway, which is a problem with streaming. The only method I could think of is MPEG-DASH streaming, as the MP4 container can store/transmit the necessary AudioSpecificConfig.
Therefore there is no simple solution to stream DAB+ services at the moment. Though I don't exclude that there exists another way of streaming where that config can be included and of which I don't know (yet). Any hints are welcome!

Instead of streaming the desired service, you could instead stream the complete ensemble by invoking `dab2eti` directly and forwarding the output via `nc`. As you use LAN, this approach does not really occupy the available network bandwidth (ETI takes ~2 MBit/s).

In your case, this should work:
Server: `dab2eti 194064000 | nc player_address 8554`
Client: `nc -l 8554 | dablin -s 0xD332`
(of course you could also use multicasting)

**Comment by malfino** [comment]
Unfortunately the idea with nc and dab2eti doesn't work for me, because I want to stream it to internet radio players, which I don't control on that level. For that I need just some standard audio stream via http. I looked through the source code of dablin and figured, that the stream is decoded to pcm before it is send to sdloutput.

The documention says: Call aacDecoder_DecodeFrame() which writes decoded PCM audio data to a client-supplied buffer. It is the client’s responsibility to allocate a buffer which is large enough to hold this output data.

Do you think it would be possible, to write the pcm-audio stream to std-out in order to be able to process it further with other programs? Like converting and streaming it via vlc or ffmpeg?

**Comment by basicmaster** [comment]
Ah, I see.
Yes, it would be possible to output the decoded audio in PCM format to `stdout` instead of forwarding it to SDL. Of course one will then have to pay attention to the specific channel count and sample rate, as these values depend on the particular service and they therefore influence the output PCM format (dablin does not do any format conversions).
I will have a look into the code how to achieve this the best way and come back to you in a few days.

**Comment by basicmaster** [comment]
You can now use the parameter `-p` to activate PCM output to `stdout` instead of SDL output.
My cmdline just for testing (local playback) was:
`dablin -p -s 0x1234 test.eti | vlc --demux=rawaud --rawaud-channels 2 --rawaud-samplerate 48000 --rawaud-fourcc fl32  -`

**Comment by malfino** [comment]
Thank you very much. After some fiddling with vlc I got it running exactly as I hoped.

`./dablin -d ../dabtools/src/dab2eti -c 7D -s 0xD332 -p | cvlc --file-caching=5000 --network-caching=5000  --sout-keep --demux=rawaud --rawaud-channels 2 --rawaud-samplerate 48000 --rawaud-fourcc fl32 stream:///dev/stdin  --no-sout-video --sout '#transcode{acodec=mp3,ab=196,channels=2,samplerate=44100}:std{access=http,mux=mp3,dst=:8080/stream.mp3}' :no-sout-rtp-sap :no-sout-standard`

I can stream it to my squeezelite-players with lms. Huzzah!

It's reencoding the stream it mp3 which is lossy. I gave it a go with flac, but that was hogging my network. The sound quality is still excellent. I wonder how it compares to the usual internet streams.
